### PR TITLE
Allow coreclr_create_delegate to return pointers to NativeCallable methods

### LINKED ISSUE #34266: Allow coreclr_create_delegate to return pointers to NativeCallable methods
Also, use it for methods that are implemented in the runtime and target of `coreclr_create_delegate`.

This will allow the methods to AOTed among other things.

### PR DESCRIPTION
**Pull request by jkotas**
Fixes #34266

### CODE REVIEW

**Review by jkoritzinsky — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
@jkotas The failures in the Windows leg are for IJW:

```
Interop\IJW\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle.cmd [FAIL]
      Fatal error. Invalid Program: attempted to call a NativeCallable method from runtime-typesafe code.
      
      Return code:      1
      Raw output file:      C:\h\w\AEC9097F\w\9E78092C\e\Interop\IJW\Reports\Interop.IJW\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle.output.txt
      Raw output:
      BEGIN EXECUTION
       "C:\h\w\AEC9097F\p\corerun.exe" LoadIjwFromModuleHandle.dll 
      Verify that we can load an IJW assembly from native code.
      Expected: 100
      Actual: -2146233082
      END EXECUTION - FAILED
      FAILED
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=C:\h\w\AEC9097F\p
      > C:\h\w\AEC9097F\w\9E78092C\e\Interop\IJW\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle.cmd
      Expected: True
      Actual:   False
      Stack Trace:
        F:\workspace\_work\1\s\artifacts\tests\coreclr\Windows_NT.x64.Checked\TestWrappers\Interop.IJW\Interop.IJW.XUnitWrapper.cs(334,0): at Interop_IJW._LoadIjwFromModuleHandle_LoadIjwFromModuleHandle_LoadIjwFromModuleHandle_._LoadIjwFromModuleHandle_LoadIjwFromModuleHandle_LoadIjwFromModuleHandle_cmd()
      Output:
        Fatal error. Invalid Program: attempted to call a NativeCallable method from runtime-typesafe code.
        
        Return code:      1
        Raw output file:      C:\h\w\AEC9097F\w\9E78092C\e\Interop\IJW\Reports\Interop.IJW\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle.output.txt
        Raw output:
        BEGIN EXECUTION
         "C:\h\w\AEC9097F\p\corerun.exe" LoadIjwFromModuleHandle.dll 
        Verify that we can load an IJW assembly from native code.
        Expected: 100
        Actual: -2146233082
        END EXECUTION - FAILED
        FAILED
        Test Harness Exitcode is : 1
        To run the test:
        > set CORE_ROOT=C:\h\w\AEC9097F\p
        > C:\h\w\AEC9097F\w\9E78092C\e\Interop\IJW\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle\LoadIjwFromModuleHandle.cmd
```

File: src/coreclr/src/vm/corhost.cpp [hunk 1]
```diff
@@ -819,11 +796,25 @@ HRESULT CorHost2::_CreateDelegate(
...
-        if (pMD==NULL || !pMD->IsStatic() || pMD->ContainsGenericVariables())
+        if (pMD==NULL || !pMD->IsStatic() || pMD->HasClassOrMethodInstantiation())
             ThrowHR(COR_E_MISSINGMETHOD);
 
-        UMEntryThunk *pUMEntryThunk = pMD->GetLoaderAllocator()->GetUMEntryThunkCache()->GetUMEntryThunk(pMD);
-        *fnPtr = (INT_PTR)pUMEntryThunk->GetCode();
+        if (pMD->HasNativeCallableAttribute())
+        {
+            if (NDirect::MarshalingRequired(pMD, pMD->GetSig(), pMD->GetModule()))
+                ThrowHR(COR_E_INVALIDPROGRAM);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/src/vm/corhost.cpp]: This API returns an `HRESULT`, is there a reason we are throwing?

Comment (jkotas) [reply src/coreclr/src/vm/corhost.cpp]: No particular reason. I am just following the style of the existing code.

**Comment by jkotas** [comment]
> LoadIjwFromModuleHandle

Tests have a direct call to this method. I have reverted NativeCallable on the IJW helper.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**